File: rust/src/leetcode/Q973_k_closest_points_to_origin.rs
/*
 * @lc app=leetcode id=973 lang=rust
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
use std::collections::BinaryHeap;

impl Solution {
    pub fn k_closest(points: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let k: usize = k as usize;
        let mut heap: BinaryHeap<(i32, (i32, i32))> = BinaryHeap::new();

        for (index, point) in points.iter().enumerate() {
            let x = point[0];
            let y = point[1];
            let distance = x * x + y * y;

            heap.push((distance, (x, y)));

            if index >= k {
                heap.pop();
            }
        }

        heap.into_iter().map(|(_, (x, y))| vec![x, y]).collect()
    }
}
// @lc code=end

File: python/Q973_k_closest_points_to_origin.py
#
# @lc app=leetcode id=973 lang=python3
#
# [973] K Closest Points to Origin
#


# @lc code=start
import heapq


class Solution:
    def kClosest(self, points: list[list[int]], k: int) -> list[list[int]]:
        min_heap: list[tuple[int, int, int]] = []

        for x, y in points:
            heapq.heappush(min_heap, (x**2 + y**2, x, y))

        result: list[list[int]] = []

        for _ in range(k):
            _, x, y = heapq.heappop(min_heap)
            result.append([x, y])

        return result


# @lc code=end
